examples/copy.rs

```diff
@@ -47,34 +47,62 @@ impl SimpleQueryHandler for DummyProcessor {
 
 #[async_trait]
 impl CopyHandler for DummyProcessor {
-    async fn on_copy_data<C>(&self, _client: &mut C, copy_data: CopyData) -> PgWireResult<()>
+    async fn on_copy_data<C>(&self, client: &mut C, copy_data: CopyData) -> PgWireResult<()>
     where
         C: ClientInfo + Sink<PgWireBackendMessage> + Unpin + Send + Sync,
         C::Error: Debug,
         PgWireError: From<<C as Sink<PgWireBackendMessage>>::Error>,
     {
+        use PgWireConnectionState::*;
+        // This is set by the `on_query` implementations while handling a
+        // `CopyIn`/`CopyOut`/`CopyBoth` response.
+        assert!(matches!(client.state(), CopyInProgress));
+
         println!("receiving data: {:?}", copy_data);
+
         Ok(())
     }
 
-    async fn on_copy_done<C>(&self, _client: &mut C, _done: CopyDone) -> PgWireResult<()>
+    async fn on_copy_done<C>(&self, client: &mut C, _done: CopyDone) -> PgWireResult<()>
     where
         C: ClientInfo + Sink<PgWireBackendMessage> + Unpin + Send + Sync,
         C::Error: Debug,
         PgWireError: From<<C as Sink<PgWireBackendMessage>>::Error>,
     {
+        use PgWireConnectionState::*;
+        // This is set by the `on_query` implementations while handling a
+        // `CopyIn`/`CopyOut`/`CopyBoth` response.
+        assert!(matches!(client.state(), CopyInProgress));
+
         println!("copy done");
+
+        client.set_state(ReadyForQuery);
         Ok(())
     }
 
-    async fn on_copy_fail<C>(&self, _client: &mut C, fail: CopyFail) -> PgWireResult<()>
+    async fn on_copy_fail<C>(&self, client: &mut C, fail: CopyFail) -> PgWireResult<()>
     where
         C: ClientInfo + Sink<PgWireBackendMessage> + Unpin + Send + Sync,
         C::Error: Debug,
         PgWireError: From<<C as Sink<PgWireBackendMessage>>::Error>,
     {
+        use PgWireConnectionState::*;
+        // This is set by the `on_query` implementations while handling a
+        // `CopyIn`/`CopyOut`/`CopyBoth` response.
+        assert!(matches!(client.state(), CopyInProgress));
+
         println!("copy failed: {:?}", fail);
-        Ok(())
+
+        // We should ensure that exit the CopyInProgress state before returning
+        // an error in order resume normal operation (in an extended protocol
+        // mode this happens only after a Sync, but the `proccess_message`
+        // handler takes care of that).
+        client.set_state(ReadyForQuery);
+        Err(PgWireError::UserError(Box::new(ErrorInfo::new(
+            "ERROR".to_owned(),
+            "XX000".to_owned(),
+            format!("COPY IN mode terminated by the user: {}", fail.message),
+        ))))
```

Comment:
As the docs mentioned, this `on_copy_fail` should always return an `Err` to that causes copy statement fail. I think we can change the return type of `on_copy_fail` to `PgWireError` directly.

Reply:
And we still need to track if the copy statement is from simple or extended subprotocol. Previously we judge it by last message type received on backend, however, both subprotocols uses same copy message type.

Reply:
This should be fix in the latest version.